### PR TITLE
Subcollection Mocking

### LINKED ISSUE #25: How does subcollection mocking work? Does it?
# Summary
Now that `FakeFirestore.FieldValue` works, my unit tests can progress a little bit more.  I'm now running into trouble where subcollection access matters.

I have code that access the database like so:
```
const snap = await firestore
  .collection("accounts")
  .doc(accountId)
  .collection("userPermissions")
  .doc(userId)
  .get();
```
This returns a snapshot whose document doesn't exist. (That is, the object returned by the mock is simply `{ exists: false }`)  This needs to not be the case.

I have my firestore mock declared like so:
```
mockFirebase({
  database: {
    accounts: [ // Root-level collection
      { // A document
        id: "testAcct",
        conversations: [ // Subcollection, right?
          { id: "testConvo", participants: ["right-user", "admin-user"] },
        ],
        userPermissions: [ // Subcollection, right??
          { id: "right-user", level: "normal" },
          { id: "admin-user", level: "admin" },
          { id: "owner", level: "super" },
        ]
      },
    ]
  }
});
```
Am I missing something here?  Is this my bad, or does the package not yet support mocking subcollections?

## Relevant information
I should note that I see no mention in the package about subcollections, so I can only assume based on the root case that this is how subcollections are meant to be represented.

To further edge the point, I threw a test case on the package, which sadly fails:
```
const db = new FakeFirestore({
  characters: [
    { id: 'homer', name: 'Homer', occupation: 'technician' },
    { id: 'krusty', name: 'Krusty', occupation: 'clown' },
    { // I added this document only
      id: 'bob',
      name: 'Bob',
      occupation: 'repairman',
      family: [ // I see this as a subcollection.  Am I wrong, or is the mock?
        { id: 'thing1', name: 'Thing 1', relation: 'pet' },
        { id: 'thing2', name: 'Thing 2', relation: 'pet' },
        { id: 'deborah', name: 'Deborah', relation: 'wife' },
      ],
    },
  ],
});

test('it can fetch records from subcollections', async () => {
  expect.assertions(2);
  const record = await db
    .collection('characters')
    .doc('bob')
    .collection('family')
    .doc('thing1')
    .get();
  expect(record.exists).toBe(true);  // Expected: true, Received: false
  expect(record.data()).toHaveProperty('name', 'Thing 1');
});
```

### Environment (if relevant)
firestore-jest-mock 0.3.0

### PR DESCRIPTION
**Pull request by AverageHelper**
# Description

Here we add full support for subcollection mocking.  To do this, we needed to restructure things a bit, change what exactly our mocked functions return, etc. This way, we no longer get `TypeError`s when trying to call typical Firestore API under our mock.

See the new paragraphs under the README for the syntax of this feature.

## Related issues

Fixes #25 

## How to test

```shell
jest
```

I've already tested this fork in a production project, and we're happy enough with it.  There are a couple more ways I'd like to be able to assert Firestore accesses, but this will do for now.

### CODE REVIEW

File: __tests__/mock-firestore.test.js
Comment (AverageHelper) [reply __tests__/mock-firestore.test.js:20]: 😎

**Comment by AverageHelper** [comment]
I agree it's a lot, but I have *no* idea how to split it up as-is. I'd be happy to go over it with you in some way if you'd like. I tried to break the big bits into their own files to make things easier. With my limited JavaScript experience, however, and with some of these parts depending on each other tightly to work properly (thinking of `CollectionReference` and `DocumentReference`), I don't know how to break it up any further. 😕

**Comment by sbatson5** [comment]
👋  @AverageHelper 

I'm happy to spitball how we can break this up. Not only will it make reviewing easier, but it will be better for any future contributors who want to look for when a particular feature was added.

The first thing that jumps out at me are the changes to `auth`:
 - `getUser` doesn't depend on anything
 - `setCustomUserClaims` doesn't depend on anything
That's an easy PR to separate.

Next, I think you could move the Query logic to its own PR without breaking much.  The only thing it depends on is `buildQuerySnapShot`, so it could be a good opportunity to move that and `buildDocFromHash` to helpers in that PR.
- Extract Query methods to Query Class
- Ensure query tests still pass
- Create helpers
- Ensure mock functions are still exported through FakeFirestore to ensure we don't introduce a breaking change (we can leave a note to add deprecation warnings for a future release).

After that, `Transaction` feels pretty isolated as well (it has no dependencies and you have it well tested).

Then, the changes you made to `Timestamp` could be a separate PR. Followed by `FieldValue`, which depends on it.

After that, I think `CollectionReference` could be its own PR as it depends on the `Query` and a few other things.  The only hurdle is not using the `DocumentReference`, so that might have to be in the same PR. This would probably be the biggest change, but it would still be much easier to parse.

In terms of digging through your commits, I wouldn't worry about maintaining that history.  I squash the commits when a PR gets merged to maintain a cleaner history (if you dig through our commit history, you'll see each commit ties to a concise PR/feature).  So, I would simply copy and paste the files in question.  The history of when certain lines were changed during development isn't as important as when they were merged.

**Comment by AverageHelper** [comment]
Will do.

**Comment by AverageHelper** [comment]
> The first thing that jumps out at me are the changes to `auth`:
> 
> * `getUser` doesn't depend on anything
> * `setCustomUserClaims` doesn't depend on anything

@sbatson5 
See #36

**Comment by AverageHelper** [comment]
I've got most of the minor changes done in #36, #37, #38, #39, and #40, in that order. I think I should probably wait until those are all merged and ready before I do the whole subcollection support bit. That makes it easier on me in case there are any changes y'all want on one of these smaller PRs.

While subcollection support would be a fairly major change, the API surface shouldn't change for those who don't rely on `FakeFirestore` having a few more methods than canon `firestore`. We'd have a few more files and more granular access to certain mock structures, but most everything still exports publicly via `FakeFirestore`, following Firestore's convention.

**Review by sbatson5 — APPROVED** [review]
Sorry for all the delays in getting this reviewed.  Even though you broke it up, it's still a pretty big PR 😅 

I think we're good to merge this, and cut a minor release, as this _shouldn't_ be a breaking change.  I'd like to get it out there for a month and if nobody reports large issues, we can do a full 1.0.0 release

**Comment by AverageHelper** [comment]
Now that we have subcollection support, I have a few ideas on how we might assert access to document paths, since assertions that `collection` or `doc` quickly become unwieldy with complex Firestore schemas. Expect an issue from me about that sometime in the next week or two.

**Pull request merged**
